AutofacTests/Ping/PingService.cs

```diff
@@ -0,0 +1,40 @@
+namespace AutofacTests.Ping
+{
+    using System;
+    using System.Configuration;
+    using System.Linq;
+    using System.ServiceModel.Configuration;
+
+    using AutofacTests.Ping.Providers;
+
+    public class PingService<TChecker, TService> : IPingService
+        where TChecker : IPingProvider
+    {
+        private readonly Func<Uri, bool> action;
+
+        private readonly Lazy<Uri> url;
+
+        public PingService(TChecker checker)
+        {
+            this.action = checker.Ping;
+            this.url = new Lazy<Uri>(() => this.GetServiceUriByServiceType(typeof(TService)));
+        }
+
+        public bool IsOnline()
+        {
+            //Console.WriteLine($"IsOnline by {typeof(TService)} pinged by [{typeof(TChecker)}]");
+            return this.action(this.url.Value);
```

Comment:
What is the reason to use deligate instead of using this.checker.Ping (remove action and add TChecker checker field)?